File: 剑指offer(Sword for offer)/用两个栈实现队列.py
# -*- coding:utf-8 -*-
class Solution:
    def __init__(self):
        self.stack1=[]
        self.stack2=[]
    def push(self, node):
        # write code here
        self.stack1.append(node)
    def pop(self):
        # return xx
        if self.stack2==[]:
            while self.stack1:
                self.stack2.append(self.stack1.pop())
            return self.stack2.pop()
        return self.stack2.pop() # python的list本身就支持pop这个函数: pop(self, index=-1, /), remove and return at index (default last)

File: 剑指offer(Sword for offer)/用两个栈实现队列.cpp
// 用两个栈来实现一个队列，完成队列的Push和Pop操作。 队列中的元素为int类型。
#include<queue>
#include<stack>
#include<iostream>
using namespace std;

// 思路：push操作：栈1专门用来接收push进来的元素。
//       pop操作：当栈2为空，将栈1的元素都push到栈2中，再从栈2里pop出去；当栈2不为空，直接从栈2pop出去。
class Solution
{
public:
	void push(int node) {
		stack1.push(node);
	}

	int pop() {
		if(stack1.empty() && stack2.empty()){return -1;}
		else{
			if(stack2.empty()){
				while(!stack1.empty()){
					int tmp = stack1.top();
					stack1.pop();
					stack2.push(tmp);
				}
				int tmp = stack2.top();
				stack2.pop();
				return tmp;
			}else{
				int tmp = stack2.top();
				stack2.pop();
				return tmp;
			}
		}
	}

	bool empty() {
		if(stack1.empty() && stack2.empty()){return true;}
		else{return false;}
	}

private:
	stack<int> stack1;
	stack<int> stack2;
};
